x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationDenialInfo.java

```diff
@@ -0,0 +1,157 @@
+    String apiKeyId,
+    String[] roles,
+    String context,
+    Collection<String> grantingClusterPrivileges,
+    Collection<String> grantingIndexPrivileges
+) {
+
+    static Builder builder(Authentication authentication, String action) {
+        return new Builder(authentication, action);
+    }
+
+    String toMessage() {
+        String userText = (isAuthenticatedWithServiceAccount ? "service account" : "user") + " [" + userPrincipal + "]";
+
+        if (runAsUserPrincipal != null) {
+            userText = userText + " run as [" + runAsUserPrincipal + "]";
+        }
+
+        if (apiKeyId != null) {
+            userText = "API key id [" + apiKeyId + "] of " + userText;
+        }
+
+        if (roles != null) {
+            userText = userText + " with roles [" + Strings.arrayToCommaDelimitedString(roles) + "]";
+        }
+
+        String message = "action [" + action + "] is unauthorized for " + userText;
+
+        if (context != null) {
+            message = message + " " + context;
+        }
+
+        if (grantingClusterPrivileges != null) {
+            message = message
+                + ", this action is granted by the cluster privileges ["
+                + collectionToCommaDelimitedString(grantingClusterPrivileges)
+                + "]";
+        }
+
+        if (grantingIndexPrivileges != null) {
+            message = message
+                + ", this action is granted by the index privileges ["
+                + collectionToCommaDelimitedString(grantingIndexPrivileges)
+                + "]";
+        }
+
+        return message;
+    }
+
+    static class Builder {
+        private final Authentication authentication;
+        private final String userPrincipal;
+        private final Boolean isAuthenticatedWithServiceAccount;
+        private final String action;
+        private String runAsUserPrincipal;
+        private String apiKeyId;
+        private String[] roles;
+        private String context;
+        private Collection<String> grantingClusterPrivileges;
+        private Collection<String> grantingIndexPrivileges;
+
+        Builder(Authentication authentication, String action) {
+            this.authentication = authentication;
+            this.action = action;
+            this.userPrincipal = authentication.getUser().authenticatedUser().principal();
+            this.isAuthenticatedWithServiceAccount = authentication.isAuthenticatedWithServiceAccount();
+        }
+
+        AuthorizationDenialInfo build() {
+            return new AuthorizationDenialInfo(
+                userPrincipal,
+                isAuthenticatedWithServiceAccount,
+                action,
+                runAsUserPrincipal,
+                apiKeyId,
+                roles,
+                context,
+                grantingClusterPrivileges,
+                grantingIndexPrivileges
+            );
+        }
+
+        public Builder withRunAsUserPrincipal() {
+            User user = authentication.getUser();
```

Comment:
The bodies of these methods are copy&paste from what used to be in `AuthorizationService`